File: src/backend/core/contracts/review.py
"""
Acceptance Criteria:
- ReviewItemResponse exposes all ReviewItem fields with typed lists parsed from JSON.
- ReviewDecision carries status (approved | rejected | edited_and_approved),
  actor_id, actor_note, and edited_content (required when status is edited_and_approved).
- ActiveICPConfigResponse exposes all ActiveICPConfig fields; JSON text lists are
  returned as list[str].
- ActiveICPConfigUpsert is the write contract for creating/replacing an active config;
  titles, geographies, signals, exclusions are list[str].
- SuppressionRuleCreate carries rule_type, value, reason, added_by.
- SuppressionRuleResponse exposes all fields.
- EnrichmentGuardrailUpsert carries guardrail_type, enabled, policy_notes, approved_by.
- EnrichmentGuardrailResponse exposes all fields.
- ConfigAuditLogResponse exposes all audit fields.
- ReviewItemCreate allows creating a review item from any source (used in tests and
  seed scripts); evidence_text must be non-empty.
- No untyped dicts as contract fields.
"""

import json
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

ReviewStatus = Literal["pending", "approved", "rejected", "edited_and_approved"]
RuleType = Literal["domain", "email", "company", "title"]
GuardrailType = Literal["enrichment_provider", "email_verification", "outreach_export"]


class ReviewItemCreate(BaseModel):
    item_type: str = Field(min_length=1)
    content: str = Field(min_length=1)
    evidence_text: str = Field(min_length=1)
    evidence_page: int | None = None
    confidence: float = Field(ge=0.0, le=1.0, default=0.0)
    source_document_id: uuid.UUID | None = None
    source_knowledge_item_id: uuid.UUID | None = None


class ReviewItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    client_id: uuid.UUID
    pipeline_id: uuid.UUID
    source_document_id: uuid.UUID | None
    source_knowledge_item_id: uuid.UUID | None
    item_type: str
    content: str
    evidence_text: str
    evidence_page: int | None
    confidence: float
    status: str
    actor_id: str | None
    actor_note: str | None
    edited_content: str | None
    decided_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ReviewDecision(BaseModel):
    status: ReviewStatus
    actor_id: str | None = None
    actor_note: str | None = None
    edited_content: str | None = None

    @model_validator(mode="after")
    def edited_content_required_when_edited(self) -> "ReviewDecision":
        if self.status == "edited_and_approved" and not self.edited_content:
            raise ValueError("edited_content is required when status is edited_and_approved")
        return self


class ActiveICPConfigUpsert(BaseModel):
    vertical: str | None = None
    target_company_size_min: int | None = None
    target_company_size_max: int | None = None
    geographies: list[str] = Field(default_factory=list)
    titles: list[str] = Field(default_factory=list)
    signals: list[str] = Field(default_factory=list)
    exclusions: list[str] = Field(default_factory=list)
    notes: str | None = None
    activated_by: str | None = None
    pipeline_config_version_id: uuid.UUID | None = None


class ActiveICPConfigResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    client_id: uuid.UUID
    pipeline_id: uuid.UUID
    pipeline_config_version_id: uuid.UUID | None
    vertical: str | None
    target_company_size_min: int | None
    target_company_size_max: int | None
    geographies: list[str]
    titles: list[str]
    signals: list[str]
    exclusions: list[str]
    notes: str | None
    activated_by: str | None
    activated_at: datetime
    created_at: datetime
    updated_at: datetime

    @field_validator("geographies", "titles", "signals", "exclusions", mode="before")
    @classmethod
    def parse_json_list(cls, v: object) -> list[str]:
        if isinstance(v, str):
            parsed = json.loads(v)
            if isinstance(parsed, list):
                return parsed
        if isinstance(v, list):
            return v
        return []


class SuppressionRuleCreate(BaseModel):
    rule_type: RuleType
    value: str = Field(min_length=1)
    reason: str | None = None
    added_by: str | None = None


class SuppressionRuleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    client_id: uuid.UUID
    pipeline_id: uuid.UUID
    rule_type: str
    value: str
    reason: str | None
    added_by: str | None
    created_at: datetime


class EnrichmentGuardrailUpsert(BaseModel):
    guardrail_type: GuardrailType
    enabled: bool
    policy_notes: str | None = None
    approved_by: str | None = None


class EnrichmentGuardrailResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    client_id: uuid.UUID
    pipeline_id: uuid.UUID
    guardrail_type: str
    enabled: bool
    policy_notes: str | None
    approved_by: str | None
    approved_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ConfigAuditLogResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    client_id: uuid.UUID
    pipeline_id: uuid.UUID
    actor_id: str | None
    action: str
    entity_type: str
    entity_id: uuid.UUID | None
    before_snapshot: str | None
    after_snapshot: str | None
    created_at: datetime

File: src/frontend/web/src/contracts/review.ts
/**
 * Acceptance Criteria:
 * - ReviewItemResponseSchema validates review item API responses.
 * - ReviewDecisionSchema validates decision payloads.
 * - ActiveICPConfigResponseSchema validates ICP config with list fields.
 * - ActiveICPConfigUpsertSchema validates write payloads.
 * - SuppressionRuleResponseSchema and SuppressionRuleCreateSchema are valid.
 * - EnrichmentGuardrailResponseSchema and EnrichmentGuardrailUpsertSchema are valid.
 * - ConfigAuditLogResponseSchema validates audit log entries.
 * - No TypeScript `any`.
 */
import { z } from "zod";

export const ReviewStatusSchema = z.enum([
  "pending",
  "approved",
  "rejected",
  "edited_and_approved",
]);

export const ReviewItemResponseSchema = z.object({
  id: z.string().min(1),
  client_id: z.string().min(1),
  pipeline_id: z.string().min(1),
  source_document_id: z.string().nullable(),
  source_knowledge_item_id: z.string().nullable(),
  item_type: z.string(),
  content: z.string(),
  evidence_text: z.string(),
  evidence_page: z.number().nullable(),
  confidence: z.number(),
  status: ReviewStatusSchema,
  actor_id: z.string().nullable(),
  actor_note: z.string().nullable(),
  edited_content: z.string().nullable(),
  decided_at: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const ReviewItemListResponseSchema = z.array(ReviewItemResponseSchema);

export const ReviewDecisionSchema = z.object({
  status: ReviewStatusSchema,
  actor_id: z.string().optional(),
  actor_note: z.string().optional(),
  edited_content: z.string().optional(),
});

export const ActiveICPConfigResponseSchema = z.object({
  id: z.string().min(1),
  client_id: z.string().min(1),
  pipeline_id: z.string().min(1),
  pipeline_config_version_id: z.string().nullable(),
  vertical: z.string().nullable(),
  target_company_size_min: z.number().nullable(),
  target_company_size_max: z.number().nullable(),
  geographies: z.array(z.string()),
  titles: z.array(z.string()),
  signals: z.array(z.string()),
  exclusions: z.array(z.string()),
  notes: z.string().nullable(),
  activated_by: z.string().nullable(),
  activated_at: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const ActiveICPConfigUpsertSchema = z.object({
  vertical: z.string().optional(),
  target_company_size_min: z.number().optional(),
  target_company_size_max: z.number().optional(),
  geographies: z.array(z.string()).default([]),
  titles: z.array(z.string()).default([]),
  signals: z.array(z.string()).default([]),
  exclusions: z.array(z.string()).default([]),
  notes: z.string().optional(),
  activated_by: z.string().optional(),
});

export const SuppressionRuleCreateSchema = z.object({
  rule_type: z.enum(["domain", "email", "company", "title"]),
  value: z.string().min(1),
  reason: z.string().optional(),
  added_by: z.string().optional(),
});

export const SuppressionRuleResponseSchema = z.object({
  id: z.string().min(1),
  client_id: z.string().min(1),
  pipeline_id: z.string().min(1),
  rule_type: z.string(),
  value: z.string(),
  reason: z.string().nullable(),
  added_by: z.string().nullable(),
  created_at: z.string(),
});

export const SuppressionRuleListResponseSchema = z.array(SuppressionRuleResponseSchema);

export const EnrichmentGuardrailUpsertSchema = z.object({
  guardrail_type: z.enum([
    "enrichment_provider",
    "email_verification",
    "outreach_export",
  ]),
  enabled: z.boolean(),
  policy_notes: z.string().optional(),
  approved_by: z.string().optional(),
});

export const EnrichmentGuardrailResponseSchema = z.object({
  id: z.string().min(1),
  client_id: z.string().min(1),
  pipeline_id: z.string().min(1),
  guardrail_type: z.string(),
  enabled: z.boolean(),
  policy_notes: z.string().nullable(),
  approved_by: z.string().nullable(),
  approved_at: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const EnrichmentGuardrailListResponseSchema = z.array(EnrichmentGuardrailResponseSchema);

export const ConfigAuditLogResponseSchema = z.object({
  id: z.string().min(1),
  client_id: z.string().min(1),
  pipeline_id: z.string().min(1),
  actor_id: z.string().nullable(),
  action: z.string(),
  entity_type: z.string(),
  entity_id: z.string().nullable(),
  before_snapshot: z.string().nullable(),
  after_snapshot: z.string().nullable(),
  created_at: z.string(),
});

export const ConfigAuditLogListResponseSchema = z.array(ConfigAuditLogResponseSchema);

export type ReviewStatus = z.infer<typeof ReviewStatusSchema>;
export type ReviewItemResponse = z.infer<typeof ReviewItemResponseSchema>;
export type ReviewDecision = z.infer<typeof ReviewDecisionSchema>;
export type ActiveICPConfigResponse = z.infer<typeof ActiveICPConfigResponseSchema>;
export type ActiveICPConfigUpsert = z.infer<typeof ActiveICPConfigUpsertSchema>;
export type SuppressionRuleCreate = z.infer<typeof SuppressionRuleCreateSchema>;
export type SuppressionRuleResponse = z.infer<typeof SuppressionRuleResponseSchema>;
export type EnrichmentGuardrailUpsert = z.infer<typeof EnrichmentGuardrailUpsertSchema>;
export type EnrichmentGuardrailResponse = z.infer<typeof EnrichmentGuardrailResponseSchema>;
export type ConfigAuditLogResponse = z.infer<typeof ConfigAuditLogResponseSchema>;
